### ISSUE TITLE
Shadows on materials with THREE.DoubleSide rendered badly in r76

### ISSUE THREAD

**Issue by williamfields** [mrdoob/three.js#8692]
##### Description of the problem 

Shadows on materials that have side:THREE.DoubleSide are rendered badly. See attached image for an example.

##### Three.js version

- [X] r76

##### Browser

- [X] Chrome

##### OS

- [X] Windows

##### Hardware Requirements (graphics card, VR Device, ...)

Unknown

![untitled](https://cloud.githubusercontent.com/assets/9380730/14693726/88e5be1e-072f-11e6-902c-791abdf88290.png)

**Comment by WestLangley** [comment]
Can you provide a test case so we can reproduce your results?

Please see "How to report a bug" in the [guidelines](https://github.com/mrdoob/three.js/blob/master/.github/CONTRIBUTING.md).

**Comment by williamfields** [comment]
Here is a test case: http://www.williamfields.com/share/threejs/bug.html

It seems to occur when both castShadow and receiveShadow are set to true on a mesh that has a double-sided material.

**Comment by tschw** [comment]
@WestLangley 
Maybe related:

We were shadowing everything single sided, which is obviously wrong for double-sided materials, so I added

https://github.com/mrdoob/three.js/blob/master/src/renderers/webgl/WebGLShadowMap.js#L368

in the clipping PR. It became very much apparent in

http://threejs.org/examples/#webgl_clipping_advanced

and obviously works fine in that example, so there must be additional factors in the mix.

**Comment by WestLangley** [comment]
@williamfields 

It appears to be self-shadowing artifacts. Try

    light.shadow.bias = - 0.001;

http://stackoverflow.com/questions/32545283/shadow-artifacts-on-double-sided-plane/32546348#32546348

**Comment by WestLangley** [comment]
@tschw 

>We were shadowing everything single sided, which is obviously wrong for double-sided materials

So back faces on double-sided materials were not receiving shadows? It seems unusual that such a fundamental issue was not previously discovered. Maybe there was some breakage during a recent refactoring?

**Comment by tschw** [comment]
@tschw
> So back faces on double-sided materials were not receiving shadows?

They were not *casting* shadows, no.

**Comment by WestLangley** [comment]
@tschw 

Yes. For depth material, or custom depth material, `material.side`  must be honored... I think it used to be... Anyway, thanks for fixing.

**Comment by williamfields** [comment]
@WestLangley 

> It appears to be self-shadowing artifacts. 

That does seem to be the issue, and setting the bias does fix it for me. Thanks!

**Comment by mrdoob** [comment]
Hmm, as far as I know, shadow maps can't work with doubleSided because it self-shadows the other side. That's why it was only back sided.

**Comment by mrdoob** [comment]
>> It appears to be self-shadowing artifacts. 

> That does seem to be the issue, and setting the bias does fix it for me. Thanks!

Hmm, so the solution is to "push" the shadow (using `bias`) a bit further?

**Comment by WestLangley** [comment]
>Hmm, so the solution is to "push" the shadow (using bias) a bit further?

Yes. And the trade-off are Peter-panning artifacts in which the shadow is slightly detached from the casting object.

**Comment by WestLangley** [comment]
>Hmm, as far as I know, shadow maps can't work with doubleSided

We spent quite a bit of time with this at one point, and as far as I know, we were able to get shadows to work properly in all instances. When artifacts were present, setting `bias` seemed to be the solution.

**Comment by williamfields** [comment]
FWIW, it seemed to work correctly in r75.

**Comment by mrdoob** [comment]
>>Hmm, as far as I know, shadow maps can't work with doubleSided

> We spent quite a bit of time with this at one point, and as far as I know, we were able to get shadows to work properly in all instances. When artifacts were present, setting `bias` seemed to be the solution.

Do you think we should change the default `bias` so it takes this into account?

**Comment by WestLangley** [comment]
>Do you think we should change the default bias so it takes this into account?

Not sure. I am inclined to think that something changed to cause this flurry of complaints. The only thing I can think of is @bhouston introduced `MeshDepthMaterial` into the shadow code -- there was some discussion of the correct depth packing logic.

Oh wait... The auto-sized shadow camera (https://github.com/mrdoob/three.js/issues/7690 and https://github.com/mrdoob/three.js/commit/b23a98f82b9493a39d52b376545de4a3d69604f4)  -- perhaps there is a loss of precision in depth because of the automatic shadow camera near and far plane settings.

**Comment by bhouston** [comment]
When this happens I use a bi-sect approach.  I check out the code at in binary search pattern until I find the commit that caused the change.  This is an important change to narrow down so I suggest that.  I unfortunately am caught on other matters for a few days so I can not help.  I do not think it is the depth packing stuff as I think it is mostly right at this point -- but it could always be.

**Comment by bhouston** [comment]
I wish we didn't have shadow bias.  Somehow Unity 3D gets away with no shadow bias and I think all users of Unity 3D greatly appreciate this.  This really should be investigated further but unfortunately I am busy until next week.

**Comment by mrdoob** [comment]
> We spent quite a bit of time with this at one point, and as far as I know, we were able to get shadows to work properly in all instances. When artifacts were present, setting `bias` seemed to be the solution.

I don't remember ever supporting DoubleSide in ShadowMap because we couldn't get it to not display the OPs output.

> We were shadowing everything single sided, which is obviously wrong for double-sided materials, so I added

> https://github.com/mrdoob/three.js/blob/master/src/renderers/webgl/WebGLShadowMap.js#L368 

Yep...

**Comment by WestLangley** [comment]
>I don't remember ever supporting DoubleSide in ShadowMap because we couldn't get it to not display the OPs output.

@mrdoob Yes, I think you are correct. Although I do not understand that sentence...

**Comment by mrdoob** [comment]
I meant that when using DoubleSide in ShadowMap we always got the glitches in the first post of this thread.

**Comment by tschw** [comment]
@williamfields 
> Here is a test case: http://www.williamfields.com/share/threejs/bug.html

Would you mind changing it to use a non-minified build (`three.js` instead of `three.min.js`), that allows the use of a debugger?

**Comment by williamfields** [comment]
@tschw 

> Would you mind changing it to use a non-minified build (three.js instead of three.min.js), that allows the use of a debugger?

Sure. It is done: http://www.williamfields.com/share/threejs/bug.html

**Comment by tschw** [comment]
Encoding `depth + exp2( 10 )` works. But that error margin is *huge* for `highp`!

I rewrote `packDepthToRGBA` to encode a float instead of a fixed point value - but still the same margin. Now looking at the decoding side...

Clues anyone? Anything `mediump` there?

**Comment by bhouston** [comment]
Getting into fights with precision never end well, and soon as you get it working on one piece of hardware it usually breaks on another. :)  I knew a guy once who could given any equation figure out the floating point error bounds perfectly and but even then it generally a sign of a bad approach if we had to get there.

I'd try to avoid this scenario completely.  Let's just not draw back faces into the depth buffer when doing shadows.

**Comment by bhouston** [comment]
Solid objects, which is what one assumes when one wants to cast shadows, should have proper front and back faces.

**Comment by tschw** [comment]
Turned out the decoding side needed `exp( -8 )`, which is at the first byte boundary. Encoding four bytes, this error is *astronomical*, thus the pack/unpack code went under "broken" in my book. 

The old routine floats around on the internet and is quite obviously the result of "fiddling around". I do not like stuff like this... It is now working correctly, needs fewer instructions and there are actually some clues in the code to understand what's going on.

@bhouston 
> Getting into fights with precision never end well, [ ... ] I'd try to avoid this scenario completely. 

Basically agree, as long as this is not interpreted as "let's not care about precision at all" and becomes a lame excuse to use inferior code.

> Let's just not draw back faces into the depth buffer when doing shadows.

Yes, that's fine. I made it switchable for now (off by default).

**Comment by tschw** [comment]
I think we can close this one. @williamfields 's code works OOTB again on current `dev`.

**Comment by mroman-ch** [comment]
Are you sure this is fixed? 

See: https://stackoverflow.com/questions/65657324/buggy-shadows-and-strange-artifacts

I found out that the issue has to do with side being Double. Sadly, setting material.side=THREE.FrontSide doesn't really fix the issue either because then objects change colors when you rotate the camera and other strange effects happen.

**Comment by mrdoob** [comment]
@mroman-ch Try to tinker a bit with [material.shadowSide](https://threejs.org/docs/#api/en/materials/Material.shadowSide).

**Comment by fynv** [comment]
Late to the party, however, an idea just came to me that an ultimate solution should be using peeling.

Set the shadow buffer value to the average of the first 2 visible layers should be fine for all cases, either double sided or not.
